### PR TITLE
fix: check GPU architecture before enabling GQA in SDPA on CUDA

### LINKED ISSUE #48633: `use_gqa_in_sdpa` doesn't check GPU architecture: up to 28% slower decode on pre-sm80 GPUs
### System Info

- transformers: 5.0.0
- torch: 2.10.0+cu128
- CUDA: 12.8
- GPU: Tesla T4 (compute capability 7.5)
- Platform: Kaggle Notebook
- Python: 3.12

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

On a T4 (sm75), SDPA always falls back to the MATH backend during decode because `use_gqa_in_sdpa()` returns `True` and `enable_gqa=True` is passed.

https://github.com/huggingface/transformers/blob/361d009f4f6b0659ceccfbeb4888df28374da2a3/src/transformers/integrations/sdpa_attention.py#L27-L36

The function only looks at `head_dim` and whether an attention mask is present — it does not check the GPU architecture. On pre-sm80 devices neither FLASH nor cuDNN kernels exist, so `enable_gqa=True` also removes EFFICIENT_ATTENTION from the candidates, leaving only MATH.

PyTorch reports both facts itself. Running the script below with `set_patch(False)` (stock behaviour) emits:

```
Flash attention only supports gpu architectures in the range [sm80, sm121]. Attempting to run on a sm 7.5 gpu.
```

```
For dense input, both fused kernels require query, key and value to have the same num_heads. Query.sizes(): [1, 12, 1, 128], Key sizes(): [1, 2, 1001, 128]
```

The first says FLASH does not exist here; the second says EFFICIENT is disqualified by the mismatched head counts that `enable_gqa` leaves in place.

### Reproduction script

Runs on a Kaggle T4 notebook. `patched()` repeats K/V before calling `sdpa_attention_forward`, which is equivalent to what an architecture check inside `use_gqa_in_sdpa` would do. Change `name` to `HuggingFaceTB/SmolLM2-360M` to reproduce on a different architecture.

```python
!pip -q install transformers accelerate
import time, torch
from transformers import AutoModelForCausalLM
import transformers.integrations.sdpa_attention as sa
from transformers.modeling_utils import ALL_ATTENTION_FUNCTIONS

# A T4 clocks down when idle. Without this warm-up the same measurement
# varies by 4-6% between runs.
def warmup_gpu(sec=1.0):
    a = torch.randn(4096, 4096, dtype=torch.float16, device="cuda")
    t_end = time.time() + sec
    while time.time() < t_end:
        a @ a
    torch.cuda.synchronize()

def pfl(ids):
    with torch.no_grad():
        return model(ids)

def dcd(prefill_out, n=32):
    past = prefill_out["past_key_values"]
    nxt = torch.argmax(prefill_out["logits"], dim=-1)[:, -1:]
    for _ in range(n):
        with torch.no_grad():
            o = model(nxt, past_key_values=past)
        past = o["past_key_values"]
        nxt = torch.argmax(o["logits"], dim=-1)

# guard against double-patching when the cell is re-run
_cur = sa.sdpa_attention_forward
if getattr(_cur, "_is_patch", False):
    _orig = sa._orig_sdpa_forward
else:
    _orig = _cur
    sa._orig_sdpa_forward = _orig

def patched(module, query, key, value, attention_mask, **kw):
    n_rep = query.shape[1] // key.shape[1]
    if n_rep > 1:
        key = key.repeat_interleave(n_rep, dim=1)
        value = value.repeat_interleave(n_rep, dim=1)
    return _orig(module, query, key, value, attention_mask, **kw)

patched._is_patch = True

def set_patch(on):
    fn = patched if on else _orig
    sa.sdpa_attention_forward = fn
    ALL_ATTENTION_FUNCTIONS["sdpa"] = fn

name = "Qwen/Qwen2.5-1.5B"
model = AutoModelForCausalLM.from_pretrained(
    name, torch_dtype=torch.float16).cuda().eval()

for L in [1000, 4000, 8000]:
    print(f"\n===== ctx={L} =====")
    for on in [False, True]:
        set_patch(on)
        ids = torch.randint(high=1000, size=[1, L], dtype=torch.int, device="cuda")
        out = pfl(ids); dcd(out, n=4)
        ts = []
        for _ in range(5):
            out = pfl(ids)
            warmup_gpu(1.0)
            torch.cuda.synchronize()
            s = torch.cuda.Event(enable_timing=True)
            e = torch.cuda.Event(enable_timing=True)
            s.record(); dcd(out, n=32); e.record()
            torch.cuda.synchronize()
            ts.append(s.elapsed_time(e) / 32)
        label = "repeat_kv" if on else "stock    "
        print(f"  {label} {min(ts):7.2f} ms  ({1000/min(ts):5.1f} tok/s)")

set_patch(False)
```

### Results

Qwen2.5-1.5B (12 Q heads / 2 KV heads)

| ctx  | stock   | with repeat_kv | latency reduction |
|------|---------|----------------|-------------------|
| 1000 | 35.72ms | 29.90ms        | 16.3%             |
| 4000 | 44.90ms | 33.77ms        | 24.8%             |
| 8000 | 71.64ms | 51.26ms        | 28.5%             |

SmolLM2-360M (15 Q heads / 5 KV heads)

| ctx  | stock   | with repeat_kv | latency reduction |
|------|---------|----------------|-------------------|
| 1000 | 38.29ms | 32.44ms        | 15.3%             |
| 4000 | 37.87ms | 31.92ms        | 15.7%             |
| 8000 | 48.78ms | 36.21ms        | 25.8%             |

Each cell is the per-token latency over 32 generated tokens, minimum of five runs. `latency reduction = (stock - patched) / stock`.

Note: the 1000 and 4000 rows are close on a T4 because the GPU is power-limited at these sizes and the extra attention work is absorbed by a clock drop; the difference becomes visible at 8000.

### Correctness

Outputs are unchanged.

- Qwen2.5-1.5B: greedy decoding produces identical token sequences for 200 tokens
- SmolLM2-360M: identical for 50 tokens
- Max absolute logit difference 0.027, mean 0.0034 (fp16)

### Expected behavior

`use_gqa_in_sdpa()` should take the GPU architecture into account and return `False` (i.e. take the `repeat_kv` path) when no fused kernel supporting GQA exists:

```python
if torch.cuda.is_available() and torch.cuda.get_device_capability()[0] < 8:
    return False
```

This check runs on every attention call, so it should probably be cached at module level rather than queried each time.

Not verified: that sm80+ devices are unaffected. Based on kernel availability this should not regress, but I don't have access to such hardware.

Related, both different symptoms:
- #45636 — head_dim > 256 and mask-layout cases; proposes the same `repeat_kv` as part of its fix
- pytorch/pytorch#139586 — notes that `enable_gqa` is only supported by the flash backend

### PR DESCRIPTION
**Pull request by truongsontung**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48635) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48635)

## Problem
On pre-sm80 GPUs (e.g. T4/sm75), FlashAttention is unavailable. When use_gqa_in_sdpa() returns True, it removes the EFFICIENT_ATTENTION candidate because the key/value head counts no longer match what the efficient kernel expects. With FlashAttention also unavailable, the only remaining backend is the slow MATH kernel, causing up to 28% slower decode on T4.

## Fix
Check CUDA compute capability in use_gqa_in_sdpa(). On pre-sm80 GPUs, return False so the EFFICIENT_ATTENTION path is preserved.

## Testing
- Added test_use_gqa_in_sdpa_pre_sm80_gpu_returns_false
- Added test_use_gqa_in_sdpa_sm80_gpu_allows_gqa

Fixes #48633

### CODE REVIEW

**Comment by truongsontung** [comment]
Hi maintainers 👋 

This PR fixes a real performance bug (28% slower decode on T4/GPU pre-sm80) reported in #48633. The CI Security Gate is blocking because it's detecting `.github` path changes that likely come from upstream divergence between this fork and the base commit.

Could someone manually approve CI? The PR only changes 2 files:
- `src/transformers/integrations/sdpa_attention.py` (fix + docstring)
- `tests/utils/test_modeling_utils.py` (2 new unit tests)

Thanks!

**Comment by banarto** [comment]
Reporter of #48633 here. I installed this branch on a Kaggle T4 (sm75) and confirmed the fix works on real hardware, with no monkey-patching:

```
pip install git+https://github.com/truongsontung/transformers.git@fix/gqa-gpu-arch-check
```

Qwen2.5-1.5B, batch 1, per-token decode latency (min of 5 runs, 32 tokens each):

| ctx  | v5.0.0  | this branch | reduction |
|------|---------|-------------|-----------|
| 1000 | 35.72ms | 31.86ms     | 10.8%     |
| 4000 | 43.84ms | 32.53ms     | 25.8%     |
| 8000 | 70.43ms | 49.31ms     | 30.0%     |

This matches what I measured with the monkey-patched version in the issue, so the fix dispatches to EFFICIENT_ATTENTION as intended.

One note in case it helps: if waiting for a manual CI approval takes a while, re-creating the branch on top of a freshly synced fork would drop the `.github` diff and let the gate pass automatically.

**Comment by banarto** [comment]
I've closed #48633 as a duplicate — #45776 covers the same issue and has been open since May, with a more general fix (`torch.backends.cuda.can_use_flash_attention()` rather than a capability threshold) and T4 measurements in [this comment](https://github.com/huggingface/transformers/pull/45776#issuecomment-5230620009).

Sorry for sending you down this path. If you're still interested in the area, #45776 is where the discussion is — it's currently blocked on a torch fullgraph-compile issue (pytorch/pytorch#182684) which was closed as completed in July, so it may be worth a re-check.

**Comment by truongsontung** [comment]
I'm aware of PR #45776 which takes a more general approach using torch.backends.cuda.can_use_flash_attention. That PR is blocked on a torch fullgraph-compile issue which was closed as completed in July.

My approach here is intentionally simpler - a capability threshold (major < 8) that avoids the torch API dependency. This may work as a targeted fix while #45776 is being unblocked.

Happy to close this in favor of #45776 if that's preferred, or adapt to use the more general approach.
